### ISSUE TITLE
[UseDotNetV2] Not setting "no_proxy" nuget config on Windows agents

### ISSUE THREAD

**Issue by gigi81** [microsoft/azure-pipelines-tasks#16935]
Entering this information will route you directly to the right team and expedite traction.

**Question, Bug, or Feature?**  
*Type*: Bug

**Enter Task Name**: UseDotNet

## Environment
- Server - Azure Pipelines or TFS on-premises? Azure Pipelines
- Agent - Hosted or Private: Private

## Issue Description

When UseDotNet (V2) runs on a Windows agent, it runs a series of nuget.exe commands at the end to setup the proxy configuration. However these series of commands do not set the "**no_proxy**" parameter resulting in a wrong configuration and following dotnet restore step failing.

### Task logs

Setting up proxy configuration for NuGet.
D:\ado\agents\agent-1\_work\_tool\NuGet\4.4.1\x64\nuget.exe config -set http_proxy=http://myproxy.com:8080
D:\ado\agents\agent-1\_work\_tool\NuGet\4.4.1\x64\nuget.exe config -set http_proxy.user=undefined
D:\ado\agents\agent-1\_work\_tool\NuGet\4.4.1\x64\nuget.exe config -set http_proxy.password=undefined

### Work Around
Added this step after UseDotNet@2 to make it work.

```yaml
- task: CmdLine@2
  displayName: 'Nuget config work around'
  condition: eq( variables['Agent.OS'], 'Windows_NT' )
  inputs:
    script: 'nuget.exe config -set no_proxy=.mydomain.com'
```

### Fix
The issues is in [nugetinstaller.ts]( https://github.com/microsoft/azure-pipelines-tasks/blob/master/Tasks/UseDotNetV2/nugetinstaller.ts)
Which is not using the [proxyBypassHosts ](https://github.com/microsoft/azure-pipelines-task-lib/blob/master/node/docs/proxy.md)parameter to set the no_proxy nuget parameter.
There should be another command issued to set the **no_proxy** parameter.

**Comment by gigi81** [comment]
Issue still present

**Comment by gigi81** [comment]
Still an issue

**Comment by wkz123** [comment]
For the love of... please fix this!

**Comment by danko-d** [comment]
You cannot simply use the content of **.proxybypass** for the **no_proxy** setting of **nuget.config**, because they are different formats:
* In **.proxybypass** you use a regex, according to https://github.com/microsoft/azure-pipelines-agent/blob/master/docs/design/proxy.md
* The **no_proxy** setting in **nuget.config** is a comma-separated list of domains the bypass the proxy server. (see https://learn.microsoft.com/en-us/nuget/reference/nuget-config-file)

**Comment by mhudasch** [comment]
Big thank you @danko-d for mentioning this. 

It took me several days to investigate that:

Hosts that should be bypassed by the agent are defined according to [https://learn.microsoft.com/en-us/azure/devops/pipelines/agents/proxy?view=azure-devops&tabs=windows#specify-proxy-bypass-urls](url)
As described these expressions are in ECMA Script regex escaped format (for example **my\\.bypasshost\\.org**) so that those lines can easily used by the pipeline tasks running in node (so the expressions must be in this format or else handling by other tasks will break) 

Later those setting are string joined with comma and added to the service user nuget.config (%AppData%/NuGet/NuGet.Config) with using `nuget config set no_proxy` as suggested by @gigi81.
This is problematic for several reasons (and for a long time at that). When running a server with multiple agent services using the same service user changes to the NuGet.config file on user-level impact all paralell executing jobs by the agent services. 
The 'workaround' by @gigi81 using `nuget config set no_proxy` before restore becomes a racecondition to edit the NuGet.config file by multiple jobs and might fail randomly (and very hard to reproduce). 
It would be way better to take the agent settings and set the environment variables that are supported by nuget.exe for the best scoping of these settings.

Finally nuget.exe (since v4+) supports proxy settings in NuGet.config but only on user-level, and http_proxy / no_proxy environment variables. As you can see here [https://github.com/NuGet/NuGet.Client/blob/a32bce39889f724fbd11cfd12e946f802168b583/src/NuGet.Core/NuGet.Configuration/Proxy/ProxyCache.cs#L80](url) if there is http_proxy setting present in the users' NuGet.config file, all settings (inlcuding no_proxy) take precedence over environment variables. The previously joined no_proxy values get splitted again and put into the bypassList. Later those values are used here [https://github.com/NuGet/NuGet.Client/blob/a32bce39889f724fbd11cfd12e946f802168b583/src/NuGet.Core/NuGet.Configuration/Proxy/WebProxy.cs#L76](url). As you can see in the `WildcardToRegex` method, those values in the list are regex-escaped again which causes a 'double-escaping' turning our sample **my\\.bypasshost\\.org** to **my\\\\.bypasshost\\\\.org** which in turn never matches the incoming host of a nuget package source.

**Comment by bramborman** [comment]
+1 for being someone who spent days to investigate why my proxy bypass settings don't work on our self-hosted build agents...

Very good research @mhudasch, though I'd go even further and say the NuGet behavior around the proxy configuration is also very confusing - as you said it takes into account only the user-specific configuration file, and environment variables as a fallback, completely overriding the documented NuGet configuration priorities that, among other say that if a config file is explicitly specified in the command line, only that file is used and nothing else... Well, not for proxy settings...

I'm gonna try to follow up to see if this is already reported for NuGet and report it eventually.

But for now I think I found a workaround, though I'm a bit unsure what side effects it may cause...
I created a `.env` file in all agents directories and specified the `APPDATA` environment variable differently for each. This way parallel builds won't interfere with each other when more agents use the same service account.
Plus, I set the agent's app data folders to be in their temp directory (that is, the `_temp` directory inside agent's work directory, which defaults to `_work` in agent's root). This temp directory is cleaned automatically after each pipeline job, and as such anything set by any build won't affect any builds running after.

The content of my `.env` file is then something like:
```
APPDATA=D:\myagent\_work\_temp\AppData
```
